platform/Span.h

```diff
@@ -0,0 +1,578 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef MBED_PLATFORM_SPAN_H_
+#define MBED_PLATFORM_SPAN_H_
+
+#include <algorithm>
+#include <stddef.h>
+#include <stdint.h>
+
+#include "platform/mbed_assert.h"
+
+namespace mbed {
+
+/**
+ * Special value for the Size parameter of Span.
+ * If the type use this value then the size of the array is stored in the object
+ * at runtime.
+ */
+#define SPAN_DYNAMIC_EXTENT -1
+
+/**
+ * View to an array.
+ *
+ * Spans encapsulate the pointer to an array and its size into a single object
+ * or type; however, it does not manage the lifetime of the array viewed.
+ * You can use instances of Span to replace the traditional pair of pointer and
+ * size arguments in function calls.
+ *
+ * You can use the size member function to query the number of elements present
+ * in the array, and overloads of the subscript operator allow code using
+ * this object to access to the content of the array viewed.
+ *
+ * @note You can create Span instances with the help of the function
+ * template make_Span() and make_const_Span().
+ *
+ * @note Span<T, Size> objects can be implicitly converted to Span<T> objects
+ * where required.
+ *
+ * @tparam T type of objects held by the array.
+ * @tparam Size The size of the array viewed. The default value
+ * SPAN_DYNAMIC_SIZE  is special as it allows construction of Span objects of
+ * any size (set at runtime).
+ */
+template<typename T, ptrdiff_t Size = SPAN_DYNAMIC_EXTENT>
+struct Span {
+
+    MBED_STATIC_ASSERT(Size >= 0, "Invalid size for a Span");
+
+    /**
+     * Construct a view to an empty array.
+     *
+     * @post a call to size() will return 0, and data() will return NULL.
+     */
+    Span() : _array(NULL) { }
+
+    /**
+     * Construct a Span from a pointer to a buffer.
+     *
+     * @param array_ptr Pointer to the array data
+     * @param array_size Number of elements of T present in the array.
+     *
+     * @post a call to size() will return array_size and data() will return
+     * @p array_ptr.
+     */
+    Span(T *array_ptr, size_t array_size) :
+        _array(array_ptr) {
+        MBED_ASSERT(array_size >= (size_t) Size);
+    }
+
+    /**
+     * Construct a Span from the reference to an array.
+     *
+     * @param elements Reference to the array viewed.
+     *
+     * @tparam Size Number of elements of T presents in the array.
+     *
+     * @post a call to size() will return Size, and data() will return
+     * a pointer to elements.
+     */
+    Span(T (&elements)[Size]):
+        _array(elements) { }
+
+    /**
+     * Return the size of the array viewed.
+     *
+     * @return The number of elements present in the array viewed.
+     */
+    size_t size() const
+    {
+        return _array ? Size : 0;
```

Comment:
Do you really need a static span to have the runtime option of being empty? C++ only permits empty for `Extent==0` or dynamic spans. From a compile-time correctness view, I quite like the idea of enforcing non-0 static spans to be non-null.

Reply:
I think it is valuable to not enforce non-0 static spans to be non-null as it can be use to represent the absence of values. That's useful for optional values represented by a static span. 

```c++
struct pdu_t { 
    // options have a fixed size 
    typedef Span<uint8_t, 5> option_t;

    uint8_t header;
    option_t option;
    Span<uint8_t> payload;
};

void process_pdu(const pdu_t& pdu) { 
    if (!pdu.option.empty()) { 
        // process options ...
    }
}
```

Reply:
Ta for references. I can see it's useful, much as allowing references to be null would be useful, but it is a major deviation from the standard C++ equivalent, where that syntax guarantees there are 5 `uint8_t`s.

I believe the C++ standard guys would tell you to use `optional<span>`, but then they're not too bothered about doubling memory usage by adding an extra `bool` alongside . In the absence of `optional`, and to save space, maybe you could designate the extended type `OptionalSpan` to make clear the extended semantics, although it becomes less catchy.

Reply:
If we ever have an optional type (would you be interested by one ? ), we can have a template specialization for `Span` that only stores the `Span` and look at `data()` to determine if the array is present or not. 

The issue with consistency is that depending on the stand point something can be declared consistent or not given that the language is full of inconsistencies. 

An array naturally decays to a pointer in pointer/boolean context so from that perspective it would be consistent to have the `operator bool() const` member function. At the same time, standard/gsl members consider that `span` should be treated as a regular container and none of the container exposes that operator: they do expose an `empty()` member function to test if the container is empty or not. 

----

To go back, to the initial issue you raised, in `std::span`, it is perfectly fine to store nothing in fixed size span. Should the code in this PR change to not accept empty span of non-0 size ?

Reply:
Agree that the language is utterly full of inconsistencies, and it's alarming that we see stuff like `string_view`->`span` being thrashed out in consecutive standards that are (from our point of view) 3-4 standards in the future. `string_view` being semi-obsolete before we ever get it is just daft.

But whenever they _do_ manage to pin something down, if we're going to copy its name I think we should copy the full semantics, just to not add to the confusion.

I'm also with Herb Sutter on the importance of [distinguishing](https://herbsutter.com/2017/02/15/distinguishing-between-maybe-null-vs-never-null-is-the-important-thing/) "maybe NULL" and "not-NULL" in the language.

In part that's because I'm aware that if the language could enforce non-NULL for pointers we could get rid of hundreds of runtime checks for NULL that get inserted to cope with unit tests that try passing NULL (sometimes as their only "test"). Unit tests may not believe in argument preconditions, but if we can stop them compiling...

So when I see that the standard has left room for that distinction, I'd like to have the chance to use it. In this form, the PR leaves no nice namespace room for the alternative "non-NULL" semantics. I'd like to reserve the name `Span` for that, with the alternative being `OptionalSpan` or equivalent. Then whether this PR actually adds `Span` or `OptionalSpan` or both is up to you - whatever you think you currently need?

And `OptionalSpan` could certainly include `bool()` as `optional<span>` does.